### PR TITLE
Turbulence

### PR DESCRIPTION
**Pull request by ESeNonFossiIo**
screen shot 2016-03-03 at 21 48 27

This is a first implementation.
TODO:
0. Check the equation with a case study
1. Better preconditioner (I have no idea at the moment)
2. Compare it with another method

I think it could be merged as it is. We could improve it with smaller PR.

### CODE REVIEW

**Comment by luca-heltai** [comment]
Very nice! 

<s>Rename `cc` file with same name of `h` file.</s>

edit: no need to rename

File: include/interfaces/turbulence_k_omega.h [hunk 1]
```diff
@@ -0,0 +1,708 @@
...
+
+private:
+  /**
+   * Enable dynamic term: \f$ \partial_t u\f$.
+   */
+  bool dynamic;
+
+  /**
+   * Enable convection term: \f$ (\nabla u)u \f$.
+   * This term introduce the non-linearity of Navier Stokes Equations.
```
Comment (luca-heltai) [inline include/interfaces/turbulence_k_omega.h]: introduces

File: include/interfaces/turbulence_k_omega.h [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+   */
+  bool convection;
+
+  /**
+   * Hot to handle \f$ (\nabla u)u \f$.
```
Comment (luca-heltai) [inline include/interfaces/turbulence_k_omega.h]: Hot -> How

File: include/interfaces/turbulence_k_omega.h [hunk 3]
```diff
@@ -0,0 +1,708 @@
...
+   */
+  std::string non_linear_term;
+
+  /**
+   * Hot to handle \f$ (\nabla u)u \f$.
```
Comment (luca-heltai) [inline include/interfaces/turbulence_k_omega.h]: Hot -> How

File: include/interfaces/turbulence_k_omega.h [hunk 4]
```diff
@@ -0,0 +1,708 @@
...
+   */
+  double rho;
+
+  /**
+   * Viscosity
+   */
+  double nu;
+
+  /**
+   * Viscosity
```
Comment (luca-heltai) [inline include/interfaces/turbulence_k_omega.h]: Same as above, or different? Please document what this is...

File: include/interfaces/turbulence_k_omega.h [hunk 5]
```diff
@@ -0,0 +1,708 @@
...
+   */
+  mutable double nu_T;
+
+  /**
+   *
```
Comment (luca-heltai) [inline include/interfaces/turbulence_k_omega.h]: Please, add a documentation of each of the parameters, maybe pointing to the equation name in the paper

File: include/interfaces/turbulence_k_omega.h [hunk 6]
```diff
@@ -0,0 +1,708 @@
...
+  bool amg_w_higher_order_elements;
+
+
+  mutable shared_ptr<TrilinosWrappers::PreconditionAMG> amg_A;
+  mutable shared_ptr<TrilinosWrappers::PreconditionAMG> amg_Ap;
+  mutable shared_ptr<TrilinosWrappers::PreconditionAMG> amg_k;
+  mutable shared_ptr<TrilinosWrappers::PreconditionAMG> amg_w;
+  mutable shared_ptr<TrilinosWrappers::PreconditionJacobi> jacobi_Mp;
+};
+
```
Comment (luca-heltai) [inline include/interfaces/turbulence_k_omega.h:159]: Since you already have a cc file, move all implementation there.

Comment (luca-heltai) [inline include/interfaces/turbulence_k_omega.h:159]: Ah, no. I see. The cc file is used for the main file.

**Comment by luca-heltai** [comment]
Please, fill in the comments in the header file, and merge. Nice work!

**Comment by asartori86** [comment]
From polimi viewpoint, it would be interesting to have this comparison between different turbulence models. A straight pipe (2D, better in r,z coordinates) with a heat flux coming from one side (with a cosine shape), a given inlet velocity and compare the average outlet temperature. This is the typical case we deal with in a nuclear reactor.

**Comment by ESeNonFossiIo** [comment]
Shuld i rewrite this code using parsedAMG? It should save a lot of Lines of code... Today i would like to do this. Is it ok?

## 

Mauro

**Comment by asartori86** [comment]
> parsedAMG

I like it :)

**Pull request merged**